### PR TITLE
Introduce opening book support

### PR DESCRIPTION
**Pull request by noobpwnftw**
My study have shown that it does no harm to use an opening book instead of a search, while the latter being largely deterministic.

This is a primitive version, using a PolyGlot-like book format but with our own Zobrist hash keys, so the probing might be cheap enough to be used elsewhere in search.

The burden is on the creation of the book to provide more diversity to the openings and cover some corner cases, which is another area of research.

No functional changes without a book.

### CODE REVIEW

**Comment by noobpwnftw** [comment]
A sample 8-move book: http://www.chessdb.cn/downloads/book.tar.gz
Project used to create books automatically: https://github.com/noobpwnftw/chessdb
A fork of python-chess to operate the book: https://github.com/noobpwnftw/python-chess

**Comment by joergoster** [comment]
Please not.
SF had book support integrated a (long) while ago. It had been dropped for good reasons.
See also https://github.com/official-stockfish/Stockfish/commit/f4dcec0b942703cf0e9bad1da7e64dfa252e915d

File: src/syzygy/tbprobe.cpp [hunk 1]
```diff
@@ -90,34 +89,6 @@ constexpr Value WDL_to_value[] = {
...
-    if ((uintptr_t)addr & (alignof(T) - 1)) // Unaligned pointer (very rare)
-        std::memcpy(&v, addr, sizeof(T));
-    else
-        v = *((T*)addr);
-
-    if (LE != IsLittleEndian)
-        swap_endian(v);
-    return v;
-}
-
```
[the remark this answers is not in the document]
Comment (noobpwnftw) [reply src/syzygy/tbprobe.cpp]: Moving endianness conversion utilities to `misc.h` since they are also used in `book.c`.

**Comment by noobpwnftw** [comment]
@joergoster I'm probing the book not only at root, but also during search to prune moves that are not in book. I see it as a search accelerator, also after this many years, a paper recently published on Science doesn't seem to have good knowledge on how to use a book during testing. :)

**Comment by Hanamuke** [comment]
Can a search that uses the book be stronger that when moves are fed by the GUI? If yes doesn't that just mean that the book is not very good? Also, for engine testing, the book is often used to force the engine to play suboptimal moves

**Comment by noobpwnftw** [comment]
Book-accelerated searches can be a lot faster, they have their candidate moves narrowed down by the book. We can build such books on better hardware and/or with unlimited time, then in daily use the book provide sufficient information to "reproduce" what was found during the creation of the book.

With this perspective, and since the book cannot be perfect in every possible line, there are cases it may be weaker than local searches, so it is better to make use of the information from the book.
For example: some book record can have a "depth" field, when SF is running on a laptop, that may be the best data it can rely on, but when SF is running on some decent hardware, it may not be that useful.

**Comment by vondele** [comment]
somewhat of a naive question. An opening book used in search, and with depth info etc. sounds a lot like the current transposition table. How different would be your plan from loading a precomputed TT ? Possibly some recipe is needed to compute this TT (zeroth order approx. a long search from the rootPos).

**Comment by MichaelB7** [comment]
Removing the old book code, although well intentioned, the removal author was only contemplating SF being used under what I would call engine tournament conditions where the book is purposely control by the GUI for both engines. It totally ignored the usage by private individual who is solely using SF for training (player training  not engine) purposes where having the use of multiple books is is a feature, where one book (GUI) guides the initial play and a second, wider book ( engine) takes over once the first book is out of book.  In some respects, removing the book code seems to be more in line with a vision statement that espouses to build the strongest engine possible while leaving open the possibility for individual users to modify the open source to use SF as they see fit.  Bottom line , I would support it being added back to the code, but I can also appreciate why it was removed in the first place.  I like the fact that you are presenting an opportunity for SF to play better in engine tournaments - as we can all see the concept ( book or no book) is getting blurred with MCTS and A0 type chess engines, what’s the real practical  difference of having an engine book or engine that has  millions of positions analyzed prior to a match ? With the hardware at our disposal , there is no doubt in my mind, we could build the strongest engine book ever - Cerebellum on steroids if you will - but that’s a direction that the maintainers would have to buy into.  I hope they do.

**Comment by mcostalba** [comment]
This is more a "study" than a patch to be committed (of course).

This project may be interesting to someone, but I'd suggest for clarity to not advertise it as a PR. This could only lead to misunderstandings.

**Comment by mcostalba** [comment]
Indeed I am not sure the official SF repo is the right place for this derived project, even as a side branch, like teh cluster one. I'd suggest to keep it in your repo and grow it (build a book, test, verify (i.e. measure) you have an advantage, etc..) there.

**Comment by noobpwnftw** [comment]
How to create a book is a derived project, having the ability to probe books is not.

If your goal is to create a minimalist chess engine at all cost, then I see plenty of code can be removed and resort to crippled alternatives provided by the GUI.

A PR page here to advertise anything sounds like a horrible idea to me, and to show an advantage I could just run it unchanged.

**Comment by mcostalba** [comment]
@noobpwnftw please could you list which functions in SF could be removed and deferred to "crippled" alternatives provided by a GUI? I am a bit curious because I fail to see any.

**Comment by Atahan-Turkoglu** [comment]
The main difference between the two maintainers: M.Costalba is dour like a boss and S.Nicolet is smiling like a director of human resources... This is the reason for S.Nicolet's great success.

**Comment by vondele** [comment]
I never like when discussions in PRs get off-topic like this.

**Comment by NickolasReynolds** [comment]
noobpwnftw's book support seems interesting to me. There is of course the tending-toward-minimalist philosophy that led to book support being removed in the past, but things are somewhat different now, I think. There are contributors and resources that allow the Stockfish project to go in directions that weren't possible or practical before.

vondele and noob are step-by-step breathing life into the cluster branch, noob is working on his book project, and -- taken together -- perhaps in the not-too-distant future it will be possible to assemble the strongest chess playing entity in history for some nefarious purpose.

It seems like some harmless fun, I'm not sure why we wouldn't want to formally embrace it.

**Comment by noobpwnftw** [comment]
@vondele 
There have never been much of a discussion, it may be too complicated to have one.

But it's all cool, I just didn't feel the need to announce it everywhere when I need to do some maintenance work(OS upgrades, etc) before holiday arrives, during which I might find nobody around to fix up things in case of something go south.

**Comment by vondele** [comment]
@noobpwnftw agreed, good to hear you're still on board!

**Comment by mcostalba** [comment]
@atahan I am happy for you that you don't seem to have an idea what a real ugly boss is :-). Good for you!


Actually I am not boss of anything, I have just written my opinion in a clear but hopefully respectful and educated way. I am sorry for the reaction, it wasn't my intention to provoke. But I'd assume that if someone posts 300 lines of new code without any kind of quantitative support apart from his good will, I would expect that he is at least ready to answer few questions, and "why we should commit this?" seems to me a legitimate question.

Anyhow I am happy he stays and keep giving his important contribution to the project

**Comment by noobpwnftw** [comment]
I see this as a simple way to reuse some form of persistent data, in order not to reproduce results that are already well-known. Books work, but they cannot provide relevant information and works only if the position is directly on the board. Using standard PolyGlot format is of course redundant, because it is expensive to recalculate the hash keys, thus not possible to probe them during search, they prove no use more than a normal book would have. I call this a "book" because it is still using the same format as PolyGlot but with our own keys, so they do not need to be recalculated, which also means we can adopt a handful of tools without much effort.

For its worth, except for the obvious openings and transpositions, there are puzzles and tricky positions which SF fails to solve efficiently, one could have them pre-calculated and probed on-the-fly, in order to improve performance on not only the exact positions but also on some of their variant forms, also, for whom with limited hardware capabilities, we could have some information pre-calculated and provided directly, but since we cannot know whether our provided information this is good enough for everyone(and it is not perfect), I think it would make sense to let the engine judge its worth, if the search would produce more reliable results than the data, then the latter may be discarded. This is something not possible with a GUI.

The whole thing is about whether we are going to define such a capability or not, my opinion is that it is actually quite common among forks to have such capabilities one way or the other. And don't get me wrong that I do not value code improvements and resort to data mining, in fact, my cores on fishtest were doing data mining and they have been moved there since I joined, not the other way around.

Also, some urban legend saying I'm playing douche games, which are not true. Many holidays are approaching, my data centers will have a complete lock down, nobody will be able to touch anything there, so there are some maintenance work to do, the cores are therefor offline for the moment, and they will be back.

**Comment by 31m059** [comment]
> I like the fact that you are presenting an opportunity for SF to play better in engine tournaments - as we can all see the concept ( book or no book) is getting blurred with MCTS and A0 type chess engines, what’s the real practical difference of having an engine book or engine that has millions of positions analyzed prior to a match ? With the hardware at our disposal , there is no doubt in my mind, we could build the strongest engine book ever - Cerebellum on steroids if you will - but that’s a direction that the maintainers would have to buy into. I hope they do.

I agree with everything @MichaelB7 said but want to especially highlight this.  The state of play of engine chess has changed dramatically in recent years.  While I'm aware of the algorithmic differences, I have some doubts whether the difference between playing millions of games against yourself and storing the positions on the one hand, and playing millions of games against yourself and black-box "learning" from them on the other, is a principled or sound one...it seems that AB engines may be accepting an unnecessary and ad-hoc handicap.

I think this is an intriguing and natural extension of SF that integrates quite nicely into the larger goals of the project.  It seems quite natural to make it a side branch like `cluster`.

**Pull request closed without merging**